Feature/intolerances

s-kond opened the pull request:
fixes #6

s-kond committed:
npm i fast fuzzy

s-kond committed:
add data to set possible intolerances-filter

s-kond committed:
implement fuzzy search

- show up to 3 filtered results live
- clicking results adds them to an activeList
- results can be removed from activeList

s-kond committed:
add intolerances
from CreateGuest-form to guestArray

s-kond committed:
small fix for extremely long names

s-kond committed:
enable editing intolerances

s-kond committed:
fix bugs

- add unique keys
- EditGuest: implement function to remove already added intolerances
from the list of possible search results

s-kond committed:
update styling

s-kond committed:
more bug fixes

s-kond committed:
add autofocus

mpagels commented on src/pages/CreateGuest.js:
```diff
@@ -12,14 +20,47 @@ export default function CreateGuest({ onHandleSubmit }) {
   }
 
+  function compareArrays(userIntolerances, allPossibleIntolerances) {
+    let editedPossibleIntolerances = [];
```
- Do you net `let` here
- why do you call the variable `editied...` when it is seems to just a `filtered` array. You did not edit somehting here. But I might be wrong.

Anyway, I challenge you to try to write this function as an one-liner. Might wanna check [array filter methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter?retiredLocale=de)

This function could be exported and tested. ;-) ❤️

s-kond replied:
I try 👍

mpagels commented on src/pages/CreateGuest.js:
```diff
@@ -29,6 +70,37 @@ export default function CreateGuest({ onHandleSubmit }) {
+        <section>
+          {activeList.map((item) => (
+            <StyledActiveIntolerance
+              key={nanoid()}
```
using `nanoid` here is not a good idea. React uses these ids, to make intern some comparing, to know, how to optimal update the DOM. When, after every rendering the id changes, this could be hard for react and is very like to create some errors or unwanted behavior. I highly recommend using unique ids, that are not chaning on every render.

mpagels commented on src/pages/CreateGuest.js:
```diff
@@ -29,6 +70,37 @@ export default function CreateGuest({ onHandleSubmit }) {
+        <section>
+          {filteredIntolerance.map((item) => (
+            <StyledSearchResult
+              key={nanoid()}
```
here as well. see my comment above.

mpagels commented on src/pages/EditGuest.js:
```diff
@@ -19,11 +30,49 @@ export default function EditGuest({ guestData, onHandleEditSubmit }) {
+      }
+    }
+    setEditedIntolerances(editedPossibleIntolerances);
+  }
```
Since I saw this code in another component, you want to DRY. So export this `utiliy` function and import it, when you needed it. This has the advantage, if you change somehting in the logic, you have to change it only in one place.

Testing the function is then easy as well.

mpagels commented on src/pages/EditGuest.js:
```diff
@@ -40,24 +89,41 @@ export default function EditGuest({ guestData, onHandleEditSubmit }) {
+        <section>
+          {activeEditList.map((item) => (
+            <StyledActiveIntolerance
+              key={nanoid()}
```
`nanoid`

mpagels commented on src/pages/EditGuest.js:
```diff
@@ -40,24 +89,41 @@ export default function EditGuest({ guestData, onHandleEditSubmit }) {
+        <section>
+          {filteredIntolerance.map((item) => (
+            <StyledSearchResult
+              key={nanoid()}
```
`nanoid`

mpagels commented on src/pages/GuestDetails.js:
```diff
@@ -21,6 +22,12 @@ export default function GuestDetails({ onDelete }) {
+        <h3>Food should be:</h3>
+        <StyledList>
+          {intolerances.map((item) => (
+            <li key={nanoid()}>{item}</li>
```
`nanoid`

mpagels reviewed (changes requested):
Hey Stefan, very awesome work here. A lot of interesting logic here.

I found some issues regardind `nanoid` as a key prop. I think this is not a good idea. See more in the comments. I left a challenge as well ;-) And some small improvements. Pls review.

s-kond committed:
improved keys

- add id to intolerances
- implement id's as keys
- changed logic to fit the new intolerances-objects

s-kond committed:
outsource & optimize compareArray-function

s-kond committed:
removed console.logs

mpagels commented on src/util/CompareArrays.js:
```diff
@@ -0,0 +1,10 @@
+//a function to compare two arrays with objects ->
+//array1 (userIntolerances) containing objects that should be deleted from array2 (allPossibleIntolerances)
```
I really like that you want to explain what the function is doing. Try this `best practise` format instead:

```js
/**
 * Returns a filtered array.
 *
 * @param {array} userIntolerances An array with user intorlerances.
 * @param {array} allPossibleIntolerances Another array with infos in it.
 * @return {array}  a filtered array.
 */

...code
```

You have to put in real information, I was just tinkering around.

mpagels replied:
aaaaaaaaaaaaaaaaaaaaand good solution. Is it working? can you write tests and check some edge cases?

s-kond replied:
ok thanks :) 
I tested it manually and it works - but I'll add a test too 👍

mpagels reviewed (approved):
Very good improvement here, Stefan!

s-kond committed:
small fixes

s-kond committed:
refactor search-intolerances as a component

s-kond committed:
add test for compare-arrays-function
